### PR TITLE
Add service layer

### PR DESCRIPTION
**Pull request by LordRaydenMK**

### CODE REVIEW

File: src/main/kotlin/io/arrowkt/service/AccountService.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    fun close(no: String, closeDate: Option<LocalDate>): AccountOperation<Account>
+
+    fun debit(no: String, amount: Amount): AccountOperation<Account>
+
+    fun credit(no: String, amount: Amount): AccountOperation<Account>
+
+    fun balance(no: String): AccountOperation<Balance>
+
+    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> = TODO()
```
Comment (LordRaydenMK) [inline src/main/kotlin/io/arrowkt/service/AccountService.kt]: I could not figure out how to use `binding` here... 

The result should be:
```
def transfer(from: String, to: String, amount: Amount): AccountOperation[(Account, Account)] = for { 
    a <- debit(from, amount)
    b <- credit(to, amount)
  } yield ((a, b))
```

Comment (pakoito) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: Oh god monster monad stack. I believe 

`typealias AccountOperation<A> = Kleisli<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository, A>`

That's what's supposed to be, and then you can get `ReaderT.monad(EitherT.monad(IO.monad()))` and that may get you `binding`.

Comment (LordRaydenMK) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: Time for me to learn Haskell....

Comment (pakoito) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: Nah we've moved past this. This builds up to something better.

Comment (LordRaydenMK) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: if I do
`typealias AccountOperation<A> = Kleisli<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository, A>`

and

```
    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> =
            binding<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository, Pair<Account, Account>>(ReaderT.monad<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository>(EitherT.monad(IO.monad()))) {
                val a = debit(from, amount).bind()
                val b = credit(to, amount).bind()
                Pair(a, b)
            }.fix()
```

I get
`/home/coka/IdeaProjects/frdomain.kt/src/main/kotlin/io/arrowkt/service/AccountService.kt: (42, 114): Type mismatch: inferred type is KleisliMonad<EitherTPartialOf<ForIO, AccountServiceException> /* = Kind<Kind<ForEitherT, ForIO>, AccountServiceException> */, AccountRepository> but Monad<EitherTPartialOf<ForIO, AccountServiceException> /* = Kind<Kind<ForEitherT, ForIO>, AccountServiceException> */> was expected`

Comment (pakoito) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: start with

```
ReaderT.monad<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository>(EitherT.monad(IO.monad())).binding {

}
```

and then see what's telling you about `debit` and `credit`. You're in the right direction, just keep on pushing!

Comment (LordRaydenMK) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: thanks @pakoito it compiles now.... and I have no idea how

File: src/main/kotlin/io/arrowkt/service/AccountService.kt [hunk 2]
```diff
@@ -27,5 +37,10 @@ interface AccountService<Account, Amount, Balance> {
 
     fun balance(no: String): AccountOperation<Balance>
 
-    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> = TODO()
+    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> =
+        ReaderT.monad<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository>(EitherT.monad(IO.monad())).binding {
```
Comment (pakoito) [inline src/main/kotlin/io/arrowkt/service/AccountService.kt:41]: 🎉

File: src/main/kotlin/io/arrowkt/service/AccountService.kt (hunk 1 above)
Comment (pakoito) [reply src/main/kotlin/io/arrowkt/service/AccountService.kt]: You were trying to pull a top-level function and IJ failed to decide which one. None were what you wanted, you needed to create the monad transformer in-place and then run it.

File: src/main/kotlin/io/arrowkt/service/AccountService.kt [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    fun close(no: String, closeDate: Option<LocalDate>): AccountOperation<Account>
+
+    fun debit(no: String, amount: Amount): AccountOperation<Account>
+
+    fun credit(no: String, amount: Amount): AccountOperation<Account>
+
+    fun balance(no: String): AccountOperation<Balance>
+
+    fun transfer(from: String, to: String, amount: Amount): AccountOperation<Pair<Account, Account>> =
+        ReaderT.monad<EitherTPartialOf<ForIO, AccountServiceException>, AccountRepository>(EitherT.monad(IO.monad())).binding {
```
Comment (pakoito) [inline src/main/kotlin/io/arrowkt/service/AccountService.kt:41]: One suggestion: move this monad instance to a variable that lives by `AccountOperation `

**Pull request merged**